include/sys/abd.h

```diff
@@ -0,0 +1,222 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2014 by Chunwei Chen. All rights reserved.
+ */
+
+/*
+ * ABD - ARC buffer data
+ * ABD is an abstract data structure for ARC. There are two types of ABD:
+ * linear for metadata and scatter for data.
+ * Their type is determined by the lowest bit of abd_t pointer.
+ * The public API will automatically determine the type
+ */
+
+#ifndef _ABD_H
+#define	_ABD_H
+
+#include <sys/zfs_context.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define	ARC_BUF_DATA_MAGIC 0xa7cb0fda
+
+#if defined(ZFS_DEBUG) && !defined(_KERNEL)
+#define	DEBUG_ABD
+#endif
+
+typedef struct arc_buf_data {
+#ifdef DEBUG_ABD
+	char		pad[PAGE_SIZE];	/* debug, coredumps when accessed */
```

Comment:
same DEBUG build error, originating from #2129 

```
error: variably modified 'pad' at file scope
```

not sure in what (best) way to fix it - so leaving it to author